lib/feature/one_hot_encoder.cc

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2021 The Clink Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fstream"
+#include "sys/stat.h"
+#include <dirent.h>
+#include <string>
+
+#include "clink/feature/one_hot_encoder.h"
+#include "nlohmann/json.hpp"
+#include "tfrt/support/error_util.h"
+
+using namespace clink;
+using namespace std;
+using namespace nlohmann;
+
+llvm::Expected<SparseVector>
+OneHotEncoderModel::transform(const int value, const int columnIndex) {
+  if (columnIndex >= model_data_.featuresizes_size()) {
+    return tfrt::MakeStringError("Column index out of range.");
+  }
+
+  int len = model_data_.featuresizes(columnIndex);
+  if (value >= len) {
+    return tfrt::MakeStringError("Value out of range.");
+  }
+  if (getDropLast()) {
+    len -= 1;
+  }
+
+  SparseVector vector(len);
+  if (value < len) {
+    vector.set(value, 1.0);
+  }
+  return vector;
+}
+
+void OneHotEncoderModel::setDropLast(const bool is_droplast) {
+  params_.is_droplast = is_droplast;
+}
+
+bool OneHotEncoderModel::getDropLast() { return params_.is_droplast; }
+
+llvm::Error OneHotEncoderModel::setModelData(const std::string model_data_str) {
+  OneHotEncoderModelDataProto model_data;
+
+  if (!model_data.ParseFromString(model_data_str)) {
+    return tfrt::MakeStringError("Failed to parse modeldata.");
+  }
+
+  for (int i = 0; i < model_data.featuresizes_size(); i++) {
+    if (model_data.featuresizes(i) <= 0) {
+      return tfrt::MakeStringError(
+          "Model data feature size value must be positive.");
+    }
+  }
+
+  model_data_ = std::move(model_data);
+
+  return llvm::Error::success();
+}
+
+/*
+ * Given a directory path, gets any one file in the directory.
+ *
+ * Flink ML saves model data in a file whose name is unknown to C++. This
+ * function helps to locat that file.
+ */
+std::string getRandomFileInDirectory(std::string path) {
```

Comment:
nits: Could we move this method to an anonymous namespace in this file?

Reply:
OK. I'll make the change.

Reply:
I'll also add more document about its behavior when there are zero or more than one files in the directory. Besides, I propose to rename this method to `getOnlyFileInDirectory`, corresponding to the `Iterables.getOnlyElement()` method.